### PR TITLE
docs: adds focus management common pattern docs

### PR DESCRIPTION
**Pull request by lindseywild**
Closes https://github.com/github/accessibility/issues/733.

Adds documentation for common patterns regarding focus management, how to test, etc. We can add to this doc once we have more finite resources, especially around the design of the focus indicators.

### CODE REVIEW

**Comment by khiga8** [comment]
I had misunderstood that we're adding our doc pages into `UI patterns` but it seems @inkblotty had meant a new `Common patterns` section ([[EPIC] [Documentation] Primer Accessibility Common Patterns #645](https://github.com/github/accessibility/issues/645)). Should we start that with this PR?

**Comment by lindseywild** [comment]
> I had misunderstood that we're adding our doc pages into `UI patterns` but it seems @inkblotty had meant a new `Common patterns` section ([[EPIC] [Documentation] Primer Accessibility Common Patterns #645](https://github.com/github/accessibility/issues/645)). Should we start that with this PR?

Oh whoops, I think you're right @khiga8 , I missed that. I can update this to reflect the new structure! For now maybe reviewers could just focus on the content until I make that update. Thanks for calling it out!

File: content/ui-patterns/focus-management.mdx [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+title: Focus management
+---
+
+## Overview
+
+Managing focus within a page or application is essential for users to successfully navigate, complete actions, and understand where they are within a web page or application. A focus indicator is usually seen visually as a blue outline around an interactive element when that element has focus.
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:7]:
```suggestion
Managing focus within a page or application is essential for users to successfully navigate, complete actions, and understand where they are. A focus indicator is usually seen visually as a blue outline around an interactive element when that element has focus.
```

**Review by camertron — APPROVED** [review]
This looks 🔥  to me, but I'm no expert. Maybe wait for another better-qualified review?

File: content/ui-patterns/focus-management.mdx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards.
+
+Things to keep in mind:
+
+- You can navigate to every interactive element (i.e. buttons, form fields, links)
+- You see a visual focus indicator for every element (i.e. a blue outline)
+- The tabbing order is logical based on the page structure and content
+- You are not trapped anywhere, and focus is never lost (i.e. when deleting a comment, focus is returned to a logical place)
+- If you can trigger a new element (i.e. a modal), focus is moved to that modal when opened, and returned to the page when closed
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:25]:
```suggestion
- You can navigate to every interactive element (for example: buttons, form fields, links)
- You see a visual focus indicator for every element (for example: a blue outline)
- The tabbing order is logical based on the page structure and content
- You are not trapped anywhere, and focus is never lost (for example: when deleting a comment, focus is returned to a logical place)
- If you can trigger a new element (for example: a modal), focus is moved to that modal when opened, and returned to the page when closed
```

File: content/ui-patterns/focus-management.mdx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+Not all users are able to use a mouse to navigate a web page or application. Keyboard-only and screen reader users rely on navigating and using a web page with only a keyboard, so they will be unable to perform certain actions if they cannot navigate to all elements.
+
+To meet [WCAG Guideline 2.1 (Keyboard Accessible)](https://www.w3.org/TR/WCAG21/#keyboard-accessible), all interactive elements need to be accessed and activated by a keyboard alone.
+
+## How to test focus management
+
+To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards.
+
+Things to keep in mind:
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:19]:
```suggestion
While using the feature, you should make sure that:
```

File: content/ui-patterns/focus-management.mdx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+- You can navigate to every interactive element (i.e. buttons, form fields, links)
+- You see a visual focus indicator for every element (i.e. a blue outline)
+- The tabbing order is logical based on the page structure and content
+- You are not trapped anywhere, and focus is never lost (i.e. when deleting a comment, focus is returned to a logical place)
+- If you can trigger a new element (i.e. a modal), focus is moved to that modal when opened, and returned to the page when closed
+
+## For designers
+
+Notate how the focus should move through various states of a feature, if applicable, for the engineer. Include where the focus should be when: an item is deleted, an item is added, an action triggers a page refresh, a modal is opened, a modal is closed, etc.
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:29]:
```suggestion
In your designs, annotate how the focus should move through various states of a feature, if applicable, for the user. Include where the focus should be when: an item is deleted, an item is added, an action triggers a page refresh, a modal is opened, a modal is closed, etc.
```

File: content/ui-patterns/focus-management.mdx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+Things to keep in mind:
+
+- You can navigate to every interactive element (i.e. buttons, form fields, links)
+- You see a visual focus indicator for every element (i.e. a blue outline)
+- The tabbing order is logical based on the page structure and content
+- You are not trapped anywhere, and focus is never lost (i.e. when deleting a comment, focus is returned to a logical place)
+- If you can trigger a new element (i.e. a modal), focus is moved to that modal when opened, and returned to the page when closed
+
+## For designers
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:27]: I'm thinking the "For designers" and "For engineers" headings should actually be h3 and there should be a h2 heading above them with a short intro of _what_ is for designers? Are the guidelines? Something else?

File: content/ui-patterns/focus-management.mdx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+When adding content to the page, focus should move to the first item that is added, unless a more logical spot is determined. For example: if there is a “Load more” button that reveals additional comments when pressed, the focus should move to the first comment that is newly added.
+
+### Removing content
+
+When removing content, focus should return to a logical spot. By default, if you remove an item and don't tell the focus where to go next, it will return to the `body` element and the user will need to navigate through the entire page again to get to where they were. For example, if deleting a comment within a thread, move the focus to either a “write comment” input if one exists, or the menu of the next comment. If there is no other comment, use the previous comment. This is just a suggestion, use best judgement when deciding what will be logical for a user.
+
+## Additional Resources
+
+### Good Focus Management
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:59]:
```suggestion
### Good focus management
```

File: content/ui-patterns/focus-management.mdx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+### Adding content
+
+When adding content to the page, focus should move to the first item that is added, unless a more logical spot is determined. For example: if there is a “Load more” button that reveals additional comments when pressed, the focus should move to the first comment that is newly added.
+
+### Removing content
+
+When removing content, focus should return to a logical spot. By default, if you remove an item and don't tell the focus where to go next, it will return to the `body` element and the user will need to navigate through the entire page again to get to where they were. For example, if deleting a comment within a thread, move the focus to either a “write comment” input if one exists, or the menu of the next comment. If there is no other comment, use the previous comment. This is just a suggestion, use best judgement when deciding what will be logical for a user.
+
+## Additional Resources
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:57]:
```suggestion
## Additional resources
```

File: content/ui-patterns/focus-management.mdx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+### Good Focus Management
+
+- [Improve Accessibility in Your React App By Managing Focus in Mutable Content](https://medium.com/swlh/improve-accessibility-in-your-react-app-by-managing-focus-in-mutable-content-4ddf4ed92186)
+- [Focus management and inert](https://css-tricks.com/focus-management-and-inert/)
+- [Removing Headaches from Focus Management](https://developers.google.com/web/updates/2016/03/focus-start-point)
+
+### Related WCAG Guidelines & Success Criteria:
```
Comment (yaili) [inline content/ui-patterns/focus-management.mdx:65]:
```suggestion
### Related WCAG guidelines and success criteria
```

**Review by yaili — DISMISSED** [review]
This is great, thank you writing this! 😍 

I made a bunch of suggestions that I think should be committed to the PR., but they are all just tiny tweaks or updating to follow our [writing guidelines](https://primer.style/contribute/documentation). Just one is about adding maybe an extra sentence or two. 🙏🏼 

Regarding the IA, I think this should move to the [Accessibility](https://primer.style/design/accessibility/accessibility-at-github) section for now. From my perspective, this is all accessibility related and if someone was looking for this info that is the section they would look for. If you don't think that is the right place, it should stay in UI patterns until we make decisions together about changes to the IA of these docs. Let's not create a new section until we have those discussions, please.

Thank you! ✨

**Comment by lindseywild** [comment]
@yaili thank you for the thorough review!

> I think this should move to the [Accessibility](https://primer.style/design/accessibility/accessibility-at-github) section for now.

Would we be ok with adding a "Common Patterns" section in the Accessibility area for the documentation our team is planning to write? [That was the initial thought in our Epic](https://github.com/github/accessibility/issues/645), I overlooked that and need to update this PR 😅

File: content/ui-patterns/focus-management.mdx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+## Why?
+
+Not all users are able to use a mouse to navigate a web page or application. Keyboard-only and screen reader users rely on navigating and using a web page with only a keyboard, so they will be unable to perform certain actions if they cannot navigate to all elements.
+
+To meet [WCAG Guideline 2.1 (Keyboard Accessible)](https://www.w3.org/TR/WCAG21/#keyboard-accessible), all interactive elements need to be accessed and activated by a keyboard alone.
+
+## How to test focus management
+
+To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards.
```
Comment (inkblotty) [inline content/ui-patterns/focus-management.mdx:17]:
```suggestion
To verify proper focus management, try using the feature or page with only a keyboard. Use the Tab key to navigate forward and Shift + Tab to move backwards. Some elements will use the arrow keys for keyboard navigation, such as toolbars and dropdowns.
```

**Review by inkblotty — APPROVED** [review]
One non-blocking suggestion. Also let's make that new Accessibility Common Patterns section 💪

**Review by langermank — APPROVED** [review]
Looks great!

**Pull request merged**